mbedtls/src/ssl/async_io.rs

```diff
@@ -102,121 +118,210 @@ impl WriteTracker {
 }
 
 /// Marker type for an IO implementation that implements both
-/// `tokio::io::AsyncRead` and `tokio::io::AsyncWrite`.
+/// [`tokio::io::AsyncRead`] and [`tokio::io::AsyncWrite`].
 pub enum AsyncStream {}
 
-// TODO: Add enum `AnyAsyncIo` as marker type for an IO implementation that
-// doesn't implement `tokio::io::AsyncRead` and `tokio::io::AsyncWrite`.
+/// Marker type for an IO implementation that
+/// doesn't implement [`tokio::io::AsyncRead`] and [`tokio::io::AsyncWrite`].
+pub enum AnyAsyncIo {}
 
-// TODO: Add `AsyncIo` trait for async IO that that doesn't implement
-// `tokio::io::AsyncRead` and `tokio::io::AsyncWrite`. For example:
-//     pub trait AsyncIo {
-//        async fn recv(&mut self, buf: &mut [u8]) -> Result<usize>;
-//         async fn send(&mut self, buf: &[u8]) -> Result<usize>;
-//     }
-// Could implement by using `async-trait` crate or
-// #![feature(async_fn_in_trait)] or Associated Types
+/// Async read or write bytes or packets.
+///
+/// Implementors represent a socket or file descriptor that can be read from or
+/// written to.
+///
+/// By implementing [`AsyncSend`] and [`AsyncRecv`] you can wrap any type of
+/// [`AsyncIo`] with [`Context<T>::establish_async`] to protect that
+/// communication channel with (D)TLS. That [`Context`] then also implements
+/// [`AsyncIo`] so you can use it interchangeably.
+///
+/// If you are using byte streams and are using [`tokio::io`], you don't need
+/// this trait and can rely on [`tokio::io::AsyncRead`] and
+/// [`tokio::io::AsyncWrite`] instead.
+#[async_trait]
+pub trait AsyncIo {
```

Comment:
Do we need both `trait AsyncIo` and [`trait AsyncSend` and `trait AsyncRecv`]? Why can't we just have one of these?

Do we need a macro for the implementations or can you just implement a private `fn async_recv`/`async_send` on a `Context` like we have for the synchronous versions?

Reply:
I see. I understand your main point here is reduce redundant trait and the dependency to `async-trait` crate.
So here :
It's better `trait AsyncIo` can be replaced by private functions `async_recv`/`async_send` on a `Context`

And let me try to implement `async_recv`/`async_send`  without using `async-trait`